### PR TITLE
Fix: Lost username when setting new URL with a path (IDFGH-904)

### PR DESCRIPTION
**Pull request by hongquan**
This PR is to fix a bug, in which, when calling `esp_http_client_set_url` with a path (`"/api/"`), the username and password set earlier with `esp_http_client_init()` are lost.

Test cases included. This is the test result:

```
Press ENTER to see the list of tests.
Running Input Param Tests...
Note: tcpip_adapter_init() has been called. Until next reset, TCP/IP task will periodicially allocate memory and consume CPU time.
E (1898) HTTP_CLIENT: config should have either URL or host & path
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:28:Input Param Tests:PASS
Running Get username...
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:54:Get username:PASS
Running Get password...
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:72:Get password:PASS
Running Username not lost...
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:94:Username not lost:PASS

-----------------------
4 Tests 0 Failures 0 Ignored 
OK
```

To able to test, I have to add 2 more functions: `esp_http_client_get_username` and `esp_http_client_get_password`.

### CODE REVIEW

**Comment by hongquan** [comment]
Add 1 more test case and this is the test result:

```
Press ENTER to see the list of tests.
Running Input Param Tests...
Note: tcpip_adapter_init() has been called. Until next reset, TCP/IP task will periodicially allocate memory and consume CPU time.
E (1602) HTTP_CLIENT: config should have either URL or host & path
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:28:Input Param Tests:PASS
Running Get username...
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:54:Get username:PASS
Running Get password...
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:72:Get password:PASS
Running Username not lost...
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:94:Username not lost:PASS
Running Username is reset...
D (1722) HTTP_CLIENT: New host assign = httpbin.org
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:121:Username is reset:PASS

-----------------------
5 Tests 0 Failures 0 Ignored 
OK
```

**Comment by dobairoland** [comment]
@mahavirj @david-cermak PTAL

**Comment by hongquan** [comment]
Hi @david-cermak I've just updated the code.

New test result:

```
Press ENTER to see the list of tests.
Running most_common_use...
Note: tcpip_adapter_init() has been called. Until next reset, TCP/IP task will periodicially allocate memory and consume CPU time.
E (2454) HTTP_CLIENT: config should have either URL or host & path
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:27:most_common_use:PASS
Running get_username_password...
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:53:get_username_password:PASS
Running username_not_lost...
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:82:username_not_lost:PASS
Running username_is_reset...
D (2548) HTTP_CLIENT: New host assign = httpbin.org
/home/quan/Works/Forks/esp-idf/components/esp_http_client/test/test_http_client.c:109:username_is_reset:PASS

-----------------------
4 Tests 0 Failures 0 Ignored 
OK
```

File: components/esp_http_client/esp_http_client.c [hunk 1]
```diff
@@ -297,6 +297,18 @@ esp_err_t esp_http_client_delete_header(esp_http_client_handle_t client, const c
     return http_header_delete(client->request->headers, key);
 }
 
+esp_err_t esp_http_client_get_username(esp_http_client_handle_t client, char **value)
```
Comment (mahavirj) [inline components/esp_http_client/esp_http_client.c]: @hongquan May I request one more change in the form of validating `client` and `value` fields in both newly added public APIs (accordingly APIs can return `ESP_ERR_INVALID_ARG`)?

Comment (hongquan) [reply components/esp_http_client/esp_http_client.c]: Sure!

**Comment by hongquan** [comment]
Hi @david-cermak 

I've rebased the commits to two.

**Comment by dobairoland** [comment]
Thank you @hongquan for your contribution. I'm submitting this for internal review. I'll contact you if other requests emerge. Otherwise, this PR will be closed when it gets merged into master.

Thank you Mahavir and David for the review.

**Pull request closed without merging**